### DISCUSSION TITLE
VSCode extension

### DISCUSSION THREAD

**Discussion by StevieD666** [clangd/clangd#852] [Q&A] [answered]
Hi all,

I'm trying to develop a VSCode extension that utilises the clangd LS. Basically I have a proprietary language that has embedded bits of C++ in it. Is there anyway I can get clangd to inspect a file that isn't open in the VSCode editor?
I've tried URIConverters as part of the language server client connection in VSCode, and although clangd says it's using the reported file - it doesn't! I've tried placing the file name in the compile_commands.json file, but the input file name get's removed and has only the compiler flags left.
I'm not a C++ developer BTW.

Any help really appreciated.

**Comment by HighCommander4** [comment, marked as answer]
I think you would have to write your own client-side LSP logic to do this. So, for example, rather than instantiating a `vscode-languageclient.LanguageClient`, you would perform your own connection to the server (perhaps reusing lower-level components of the [vscode client implementation](https://github.com/microsoft/vscode-languageserver-node/tree/main/client) like JSON-RPC).

Then, instead of relying on the `LanguageClient` to have registered a listener that triggers when a file is opened and send the `textDocument/didOpen` message with the entire contents of the file, you register your own listener and send the `textDocument/didOpen` message once for each piece of C++ code in the file, with just that piece as the content -- and so on for other messages.

**Reply by HighCommander4** [reply]
By the way, this use case is somewhat common, and there is an [open feature request](https://github.com/microsoft/vscode/issues/1751) for VSCode to make things like this easier. Perhaps the comments in that issue contain other suggestions for how this can be done today.
